### DISCUSSION TITLE
Imported Mode outside of Jupyter?

### DISCUSSION THREAD

**Discussion by tabreturn** [py5coding/py5generator#19] [Q&A] [answered]
@hx2A -- is there a way to use [imported mode](http://py5.ixora.io/tutorials/py5-modes/#imported-mode) outside of Jupyter? I tried, but I still need a `py5` prefix for `mouse_x`/`mouse_y` and `frame_count`, plus I have to add a `import py5` line.

```python
import py5
from py5 import *

def settings():
    size(500, 500)

def setup():
    fill(255, 0, 0)
    no_stroke()

def draw():
    circle(py5.mouse_x, py5.mouse_y, 10)
    print(py5.frame_count)

run_sketch()
```

**Comment by hx2A** [comment, marked as answer]
Yes! There is a command line utility called `run_sketch`:

```
$ run_sketch -h
usage: run_sketch [-h] [-c CLASSPATH] sketch_path

Execute py5 sketch

positional arguments:
  sketch_path           path to py5 sketch

optional arguments:
  -h, --help            show this help message and exit
  -c CLASSPATH, --classpath CLASSPATH
                        extra directories to add to classpath
```
I've been focusing most of my "imported mode" efforts into Jupyter so I haven't put as much effort into this command line tool lately. But it should work. Save the below code to a file and pass the file to that utility:

```
def settings():
    size(500, 500)

def setup():
    fill(255, 0, 0)
    no_stroke()

def draw():
    circle(mouse_x, mouse_y, 10)
    print(frame_count)
```

**Reply by hx2A** [reply]
And here's some more information about imported mode.

The Imported Mode kernel is almost identical to the regular IPython kernel. All of the code fits in one file, [kernel.py](https://github.com/hx2A/py5generator/blob/main/py5_resources/py5_module/py5_tools/kernel/kernel.py). The IPython kernel developers deserve a lot of credit for their well designed code; they thought about many possible customizations people might want and need, and made them easy to implement.

Look at the `_KERNEL_STARTUP` variable: that's the code that runs before the user enters code into a single cell. It takes care of most of the magic that makes imported mode work. There's also a necessary AST transformation that converts code like `mouse_x` to `mouse_x()`. It basically adds an extra `()` to the end of the dynamic variables.

If you wanted to mimic imported mode outside of Jupyter and without the `run_sketch` command line tool, you'd have to put the `_KERNEL_STARTUP` code at the top of your program and in your code would have to add `()` to a few places because there would be no AST transformation. Your previous example would look like this:

```
import py5_tools      
py5_tools.set_imported_mode(True)
from py5 import *

def settings():
    size(500, 500)

def setup():
    fill(255, 0, 0)
    no_stroke()

def draw():
    circle(mouse_x(), mouse_y(), 10)
    print(frame_count)

run_sketch()
```

Are you on OSX? I should have mentioned this before, but for OSX users, the `%gui osx` magic is necessary. I haven't figured out a way to get py5 to work on OSX without that.

**Reply by hx2A** [reply]
Also, after the next release, you'll be able to put the call to `size()` in `setup()` instead of needing `settings()`.

**Reply by tabreturn** [reply]
No, I'm using Linux. BTW, the `print(frame_count)` prints an object address, not an integer. So I'll stick to using run_sketch.

**Reply by hx2A** [reply]
> BTW, the `print(frame_count)` prints an object address, not an integer. So I'll stick to using run_sketch.

Wait, how are you getting an object address? I don't see that. You should be getting an integer.

**Reply by hx2A** [reply]
I made a mistake in my earlier comment. The code should have been:

```
import py5_tools      
py5_tools.set_imported_mode(True)
from py5 import *

def settings():
    size(500, 500)

def setup():
    fill(255, 0, 0)
    no_stroke()

def draw():
    circle(mouse_x(), mouse_y(), 10)
    print(frame_count())

run_sketch()
```

I forgot to add the `()` after `frame_count`.

Trying to use py5 like this is a bad practice because it leads to those kinds of errors. But I mentioned this approach so you would understand what is happening and how it works. Using `run_sketch` or the py5 kernel are much better choices, and are what I much prefer people use for their coding efforts.

**Comment by tabreturn** [comment]
Fantastic! Thanks :smile:

**Reply by tabreturn** [reply]
I just realised that `run_sketch.py` supports static-mode sketches, too! Awesome :sunglasses:

**Reply by hx2A** [reply]
When you say static mode sketches, do you mean a Sketch without any functions? For example:

```
size(200, 200)
fill(255, 0, 0)
rect(10, 20, 30, 40)
```
I forgot to mention it can do that. I agree, it is useful! It basically works by indenting the code and adding a `setup()` at the top.

You can get similar functionality in Jupyter (in either the py5 kernel + imported mode or the regular IPython kernel) using the magic commands such as [%%py5draw](http://py5.ixora.io/reference/py5draw/).

Soon I intend to make a second Jupyter kernel called "py5bot" that works kind of like drawbot or shoebot. It will pass the cell contents to the same code that backs `run_sketch.py`.
